// @polkadot-api/known-chains/dist/index.mjs
import {
  chainSpec as chainSpec16
} from "./chunk-QIV5AJQF.mjs";
import {
  chainSpec as chainSpec17
} from "./chunk-TP7HBWY6.mjs";
import {
  chainSpec as chainSpec18
} from "./chunk-URFAVCF4.mjs";
import {
  chainSpec as chainSpec19
} from "./chunk-O3YJ5EQO.mjs";
import {
  chainSpec as chainSpec20
} from "./chunk-TNXLGJUM.mjs";
import {
  chainSpec as chainSpec21
} from "./chunk-ZIBWTERM.mjs";
import {
  chainSpec as chainSpec22
} from "./chunk-YMWTNFEZ.mjs";
import {
  chainSpec as chainSpec23
} from "./chunk-QZBWUVER.mjs";
import {
  chainSpec as chainSpec8
} from "./chunk-5VZIQBMS.mjs";
import {
  chainSpec as chainSpec9
} from "./chunk-LG5IOLTV.mjs";
import {
  chainSpec as chainSpec10
} from "./chunk-JXLVOOH5.mjs";
import {
  chainSpec as chainSpec11
} from "./chunk-SOINKWG7.mjs";
import {
  chainSpec as chainSpec12
} from "./chunk-AVS4PDRR.mjs";
import {
  chainSpec as chainSpec13
} from "./chunk-WPYCR7LE.mjs";
import {
  chainSpec as chainSpec14
} from "./chunk-V5VATG3Q.mjs";
import {
  chainSpec as chainSpec15
} from "./chunk-3NBF63QS.mjs";
import {
  chainSpec
} from "./chunk-5FIWG7XZ.mjs";
import {
  chainSpec as chainSpec2
} from "./chunk-DQYTBFVY.mjs";
import {
  chainSpec as chainSpec3
} from "./chunk-QE62BN6Z.mjs";
import {
  chainSpec as chainSpec4
} from "./chunk-D7XKE5XB.mjs";
import {
  chainSpec as chainSpec5
} from "./chunk-4Q5KAEDT.mjs";
import {
  chainSpec as chainSpec6
} from "./chunk-ESSH3YTM.mjs";
import {
  chainSpec as chainSpec7
} from "./chunk-DPYHF5EQ.mjs";
export {
  chainSpec as ksmcc3,
  chainSpec2 as ksmcc3_asset_hub,
  chainSpec3 as ksmcc3_bridge_hub,
  chainSpec4 as ksmcc3_encointer,
  chainSpec5 as ksmcc3_people,
  chainSpec6 as paseo,
  chainSpec7 as paseo_asset_hub,
  chainSpec8 as paseo_people,
  chainSpec9 as polkadot,
  chainSpec10 as polkadot_asset_hub,
  chainSpec11 as polkadot_bridge_hub,
  chainSpec12 as polkadot_collectives,
  chainSpec13 as polkadot_coretime,
  chainSpec14 as polkadot_people,
  chainSpec15 as rococo_v2_2,
  chainSpec16 as rococo_v2_2_asset_hub,
  chainSpec17 as rococo_v2_2_bridge_hub,
  chainSpec18 as rococo_v2_2_people,
  chainSpec19 as westend2,
  chainSpec20 as westend2_asset_hub,
  chainSpec21 as westend2_bridge_hub,
  chainSpec22 as westend2_collectives,
  chainSpec23 as westend2_people
};
