// @subsquid/substrate-processor/lib/selection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBlockHeaderSelectionValidator = getBlockHeaderSelectionValidator;
exports.getExtrinsicSelectionValidator = getExtrinsicSelectionValidator;
exports.getEventSelectionValidator = getEventSelectionValidator;
exports.getCallSelectionValidator = getCallSelectionValidator;
exports.getFieldSelectionValidator = getFieldSelectionValidator;
const util_internal_validation_1 = require("@subsquid/util-internal-validation");
const FIELD = (0, util_internal_validation_1.option)(util_internal_validation_1.BOOLEAN);
function getBlockHeaderSelectionValidator() {
    let fields = {
        digest: FIELD,
        extrinsicsRoot: FIELD,
        stateRoot: FIELD,
        timestamp: FIELD,
        validator: FIELD,
    };
    return (0, util_internal_validation_1.object)(fields);
}
function getExtrinsicSelectionValidator() {
    let fields = {
        hash: FIELD,
        error: FIELD,
        fee: FIELD,
        signature: FIELD,
        success: FIELD,
        tip: FIELD,
        version: FIELD,
    };
    return (0, util_internal_validation_1.object)(fields);
}
function getEventSelectionValidator() {
    let fields = {
        args: FIELD,
        name: FIELD,
        phase: FIELD,
        topics: FIELD,
    };
    return (0, util_internal_validation_1.object)(fields);
}
function getCallSelectionValidator() {
    let fields = {
        args: FIELD,
        error: FIELD,
        name: FIELD,
        origin: FIELD,
        success: FIELD
    };
    return (0, util_internal_validation_1.object)(fields);
}
function getFieldSelectionValidator() {
    return (0, util_internal_validation_1.object)({
        block: (0, util_internal_validation_1.option)(getBlockHeaderSelectionValidator()),
        extrinsic: (0, util_internal_validation_1.option)(getExtrinsicSelectionValidator()),
        call: (0, util_internal_validation_1.option)(getCallSelectionValidator()),
        event: (0, util_internal_validation_1.option)(getEventSelectionValidator()),
    });
}
