// @subsquid/scale-type-system/lib/type-checker.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypeChecker = exports.BaseType = void 0;
const assert_1 = __importDefault(require("assert"));
class BaseType {
    phantom() {
        throw new Error();
    }
}
exports.BaseType = BaseType;
class TypeChecker {
    constructor(types) {
        this.types = types;
        this.checked = new Array(this.types.length);
        this.checking = new Array(this.types.length);
        for (let i = 0; i < this.checked.length; i++) {
            this.checked[i] = new WeakMap();
            this.checking[i] = new Set();
        }
    }
    getScaleType(ti) {
        (0, assert_1.default)(0 <= ti && ti < this.types.length);
        return this.types[ti];
    }
    match(ti, type) {
        (0, assert_1.default)(0 <= ti && ti < this.types.length);
        let checked = this.checked[ti];
        let ok = checked.get(type);
        if (ok == null) {
            let checking = this.checking[ti];
            if (checking.has(type))
                return true;
            checking.add(type);
            try {
                ok = type.match(this, this.getScaleType(ti));
            }
            finally {
                checking.delete(type);
            }
            checked.set(type, ok);
        }
        return ok;
    }
}
exports.TypeChecker = TypeChecker;
