// @subsquid/util-internal-validation/lib/error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataValidationError = exports.ValidationFailure = void 0;
const util_1 = require("./util");
class ValidationFailure {
    constructor(value, message) {
        this.value = value;
        this.message = message;
        this.path = [];
    }
    toString() {
        let msg = this.message;
        if (msg.includes('{value}')) {
            msg = msg.replace('{value}', (0, util_1.print)(this.value));
        }
        if (this.path.length) {
            msg = `invalid value at ${this.getPathString()}: ${msg}`;
        }
        return msg;
    }
    getPathString() {
        let s = '';
        for (let i = this.path.length - 1; i >= 0; i--) {
            s += '/' + this.path[i];
        }
        return s;
    }
}
exports.ValidationFailure = ValidationFailure;
class DataValidationError extends Error {
    get name() {
        return 'DataValidationError';
    }
}
exports.DataValidationError = DataValidationError;
