// @subsquid/substrate-data/lib/parsing/event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.decodeEvents = decodeEvents;
const sts_1 = require("@subsquid/substrate-runtime/lib/sts");
const util_1 = require("../types/util");
const EventItem = (0, sts_1.struct)({
    phase: (0, sts_1.closedEnum)({
        Initialization: (0, sts_1.unknown)(),
        ApplyExtrinsic: (0, sts_1.number)(),
        Finalization: (0, sts_1.unknown)()
    }),
    event: (0, sts_1.externalEnum)(),
    topics: (0, sts_1.array)((0, sts_1.bytes)())
});
const EventItemList = (0, sts_1.array)(EventItem);
function decodeEvents(runtime, eventsStorageValue) {
    (0, util_1.assertStorage)(runtime, 'System.Events', ['Required', 'Default'], [], EventItemList);
    let items = runtime.decodeStorageValue('System.Events', eventsStorageValue);
    return items.map((it, index) => {
        let { name, args } = runtime.toEventRecord(it.event);
        let e = {
            index,
            name,
            args,
            topics: it.topics,
            phase: it.phase.__kind
        };
        if (it.phase.__kind == 'ApplyExtrinsic') {
            e.extrinsicIndex = it.phase.value;
        }
        return e;
    });
}
