// @subsquid/util-internal-http-server/lib/path-pattern.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PathPattern = void 0;
class PathPattern {
    constructor(pattern) {
        let names = [];
        let parts = pattern.split('{');
        let re = escapeRegex(parts[0]);
        for (let part of parts.slice(1)) {
            let pair = part.split('}');
            if (pair.length != 2)
                throw new Error(`invalid pattern: ${pattern}`);
            names.push(pair[0]);
            re += '([^\\\\/]+)';
            re += escapeRegex(pair[1]);
        }
        this.regex = new RegExp('^' + re + '$');
        this.names = names;
    }
    match(path) {
        let m = this.regex.exec(path);
        if (m)
            return this.toParams(m);
    }
    toParams(m) {
        let params = {};
        for (let i = 0; i < this.names.length; i++) {
            try {
                params[this.names[i]] = decodeURIComponent(m[i + 1]);
            }
            catch (err) {
                return undefined;
            }
        }
        return params;
    }
}
exports.PathPattern = PathPattern;
function escapeRegex(s) {
    return s.replace(/([.*+?=^!:${}()|[\]\/\\])/g, '\\$1');
}
