// @subsquid/rpc-client/lib/rate.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RateMeter = void 0;
const assert_1 = __importDefault(require("assert"));
class RateMeter {
    constructor(windowSize = 10, slotTime = 100) {
        this.windowSize = windowSize;
        this.slotTime = slotTime;
        this.time = 0;
        this.window = new Array(windowSize + 1);
        this.window.fill(0);
    }
    inc(count, now) {
        now = this.toTime(now);
        let cutoff = now - this.window.length;
        if (this.time > cutoff) {
            while (cutoff > this.time - this.window.length) {
                this.window[cutoff % this.window.length] = 0;
                cutoff -= 1;
            }
        }
        else {
            this.window.fill(0);
        }
        this.window[now % this.window.length] += count;
        this.time = now;
    }
    getRate(now) {
        now = this.toTime(now);
        let cutoff = now - this.window.length;
        let time = this.time;
        let rate = 0;
        while (time > cutoff) {
            rate += this.window[time % this.window.length];
            time -= 1;
        }
        return 1000 * rate / (this.windowSize * this.slotTime);
    }
    toTime(now) {
        now = now ?? Date.now();
        let time = Math.ceil(now / this.slotTime);
        time = Math.max(time, this.time);
        (0, assert_1.default)(time > this.window.length);
        return time;
    }
}
exports.RateMeter = RateMeter;
