// @subsquid/scale-codec/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypeKind = void 0;
var TypeKind;
(function (TypeKind) {
    TypeKind[TypeKind["Primitive"] = 0] = "Primitive";
    TypeKind[TypeKind["Compact"] = 1] = "Compact";
    TypeKind[TypeKind["Sequence"] = 2] = "Sequence";
    TypeKind[TypeKind["BitSequence"] = 3] = "BitSequence";
    TypeKind[TypeKind["Array"] = 4] = "Array";
    TypeKind[TypeKind["Tuple"] = 5] = "Tuple";
    TypeKind[TypeKind["Composite"] = 6] = "Composite";
    TypeKind[TypeKind["Variant"] = 7] = "Variant";
    TypeKind[TypeKind["Option"] = 8] = "Option";
    TypeKind[TypeKind["DoNotConstruct"] = 9] = "DoNotConstruct";
    TypeKind[TypeKind["BooleanOption"] = 10] = "BooleanOption";
    TypeKind[TypeKind["Bytes"] = 11] = "Bytes";
    TypeKind[TypeKind["BytesArray"] = 12] = "BytesArray";
    TypeKind[TypeKind["HexBytes"] = 13] = "HexBytes";
    TypeKind[TypeKind["HexBytesArray"] = 14] = "HexBytesArray";
    /**
     * @internal
     */
    TypeKind[TypeKind["Struct"] = 15] = "Struct";
})(TypeKind || (exports.TypeKind = TypeKind = {}));
